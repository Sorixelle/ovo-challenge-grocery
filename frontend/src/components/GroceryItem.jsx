import React, { useState } from 'react'
import http from '../http'

function GroceryItem ({ item, quantity, category, updateItem, deleteItem }) {
  const [editMode, setEditMode] = useState(false)

  const [newQuantity, setQuantity] = useState(quantity)
  const [newCategory, setCategory] = useState(category === null ? '' : category)

  async function handleEdit () {
    const updatedItem = {
      quantity: parseInt(newQuantity)
    }
    if (newCategory !== '') {
      updatedItem.category = newCategory
    }

    const { data } = await http.patch(`/item/${item}`, updatedItem)

    updateItem(item, data)
    setEditMode(false)
  }

  async function handleDelete () {
    await http.delete(`/item/${item}`)

    deleteItem(item)
  }

  return (
    <div className='list-group-item'>
      {editMode
        ? (
          <input
            type='number' value={newQuantity}
            style={{ width: '5%' }} onChange={e => setQuantity(e.target.value)}
          />
        ) : quantity + 'x'}
      {' ' + item + ' '}
      {editMode
        ? (
          <input
            type='test' value={newCategory}
            placeholder='Category' onChange={e => setCategory(e.target.value)}
          />
        ) : <small className='text-muted'>{category}</small>}
      <a
        className={
          'btn text-white float-right ' + (editMode ? 'btn-success' : 'btn-primary')
        } onClick={() => {
          if (editMode) {
            handleEdit()
          } else {
            setEditMode(true)
          }
        }}
      >
        {editMode ? 'Save' : 'Edit'}
      </a>
      {editMode
        ? (
          <a
            className='btn btn-danger float-right text-white' onClick={handleDelete}
          >
            Delete
          </a>
        ) : ''}
    </div>
  )
}

export default GroceryItem
