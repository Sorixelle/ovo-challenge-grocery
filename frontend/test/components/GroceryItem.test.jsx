import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import GroceryItem from '../../src/components/GroceryItem'

describe('GroceryItem', () => {
  const node = <GroceryItem item='test' quantity='1' category='test' />
  it('should match the snapshot', () => {
    const { container } = render(node)
    expect(container).toMatchSnapshot()
  })

  it('should go into edit mode when the button is clicked', () => {
    const { getByText, getByDisplayValue, getByPlaceholderText } = render(node)

    const editButton = getByText('Edit')
    fireEvent.click(editButton)

    expect(editButton).toHaveClass('btn-success')
    expect(editButton).not.toHaveClass('btn-primary')

    const quantityInput = getByDisplayValue('1')
    expect(quantityInput).toBeInTheDocument()

    const categoryInput = getByPlaceholderText('Category')
    expect(categoryInput).toBeInTheDocument()
  })
})
