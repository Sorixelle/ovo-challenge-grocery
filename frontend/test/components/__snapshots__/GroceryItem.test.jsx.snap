// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroceryItem should match the snapshot 1`] = `
<div>
  <div
    class="list-group-item"
  >
    1x
     test 
    <small
      class="text-muted"
    >
      test
    </small>
    <a
      class="btn text-white float-right btn-primary"
    >
      Edit
    </a>
    
  </div>
</div>
`;
